File: src/chrome/browser/offline_items_collection/offline_content_aggregator_factory.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/offline_items_collection/offline_content_aggregator_factory.h"

#include "base/memory/singleton.h"
#include "build/build_config.h"
#include "chrome/browser/profiles/incognito_helpers.h"
#include "components/keyed_service/content/browser_context_dependency_manager.h"
#include "components/offline_items_collection/core/offline_content_aggregator.h"
#include "content/public/browser/browser_context.h"

#if defined(OS_ANDROID)
static offline_items_collection::OfflineContentAggregator*
    g_offline_content_aggregator = nullptr;
#endif

// static
OfflineContentAggregatorFactory*
OfflineContentAggregatorFactory::GetInstance() {
  return base::Singleton<OfflineContentAggregatorFactory>::get();
}

// static
offline_items_collection::OfflineContentAggregator*
OfflineContentAggregatorFactory::GetForBrowserContext(
    content::BrowserContext* context) {
#if defined(OS_ANDROID)
  if (!g_offline_content_aggregator) {
    g_offline_content_aggregator =
        new offline_items_collection::OfflineContentAggregator();
  }
  return g_offline_content_aggregator;
#else
  return static_cast<offline_items_collection::OfflineContentAggregator*>(
      GetInstance()->GetServiceForBrowserContext(context, true));
#endif
}

OfflineContentAggregatorFactory::OfflineContentAggregatorFactory()
    : BrowserContextKeyedServiceFactory(
          "offline_items_collection::OfflineContentAggregator",
          BrowserContextDependencyManager::GetInstance()) {}

OfflineContentAggregatorFactory::~OfflineContentAggregatorFactory() = default;

KeyedService* OfflineContentAggregatorFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  DCHECK(!context->IsOffTheRecord());
  return new offline_items_collection::OfflineContentAggregator();
}

content::BrowserContext*
OfflineContentAggregatorFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}

File: src/chrome/android/java/src/org/chromium/chrome/browser/download/items/OfflineContentAggregatorFactory.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.download.items;

import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;

import org.chromium.chrome.browser.ChromeFeatureList;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.components.offline_items_collection.OfflineContentProvider;

/**
 * Basic factory that creates and returns an {@link OfflineContentProvider} that is attached
 * natively to {@link Profile}.
 */
public class OfflineContentAggregatorFactory {
    // TODO(crbug.com/857543): Remove this after downloads have implemented it.
    // We need only one provider, since OfflineContentAggregator lives in the original profile.
    private static OfflineContentProvider sProvider;

    private OfflineContentAggregatorFactory() {}

    /**
     * Allows tests to push a custom {@link OfflineContentProvider} to be used instead of the one
     * pulled from a {@link Profile}.
     * @param provider The {@link OfflineContentProvider} to return.  If {@code null}, will revert
     *                 to the non-overriding behavior and pull a {link OfflineContentProvider} from
     *                 {@link Profile}.
     */
    @VisibleForTesting
    public static void setOfflineContentProviderForTests(
            @Nullable OfflineContentProvider provider) {
        if (provider == null) {
            sProvider = null;
        } else {
            sProvider = getProvider(provider);
        }
    }

    /**
     * Used to get access to the offline content aggregator.
     * @return An {@link OfflineContentProvider} instance representing the offline content
     *         aggregator.
     */
    public static OfflineContentProvider get() {
        if (sProvider == null) {
            sProvider = getProvider(nativeGetOfflineContentAggregator());
        }
        return sProvider;
    }

    private static OfflineContentProvider getProvider(OfflineContentProvider provider) {
        if (ChromeFeatureList.isEnabled(ChromeFeatureList.DOWNLOAD_OFFLINE_CONTENT_PROVIDER)) {
            return provider;
        } else {
            return new DownloadBlockedOfflineContentProvider(provider);
        }
    }

    private static native OfflineContentProvider nativeGetOfflineContentAggregator();
}
